File: C_TRR/15.c
// Viết bằng C
// Bài 15: Tính lũy thừa nhanh (Fast Exponentiation).

// CÁCH 1: ĐỆ QUY
#include <stdio.h>
 
long long fast_exponentiation_recursive(long long a, long long b)  
// Hàm đệ quy tính lũy thừa nhanh
{
    if (b == 0) return 1;  // Cơ sở: a^0 = 1
    long long half = fast_exponentiation_recursive(a, b / 2); // Gọi đệ quy với b/2 
    return (b % 2 == 0) ? (half * half) : (a * half * half); // Trả về kết quả
}

int main() 
{
    printf("%lld\n", fast_exponentiation_recursive(2, 10));  // 1024
    return 0;
}


// CÁCH 2: VÒNG LẶP
#include <stdio.h>

long long fast_exponentiation_iterative(long long a, long long b) 
// Hàm vòng lặp tính lũy thừa nhanh
{
    long long result = 1; // Khởi tạo kết quả là 1 
    while (b > 0)  // Vòng lặp với điều kiện b > 0
    {
        if (b % 2 == 1) result *= a;  // Nếu b lẻ, nhân với a
        a *= a;  // Bình phương a
        b /= 2;  // Chia b cho 2
    }
    return result;
}

int main() 
{
    printf("%lld\n", fast_exponentiation_iterative(2, 10));  // 1024
    return 0;
}

File: Python_TRR/15.py
# Viết bằng Python
# Bài 15: Tính lũy thừa nhanh (Fast Exponentiation).

# Cách 1: Sử dụng đệ quy
def fib_recursive(n): # F(n) = F(n-1) + F(n-2) với F(0) = 0, F(1) = 1
    """Tính số Fibonacci thứ n bằng đệ quy."""
    if n <= 1: # Trường hợp cơ bản F(0) = 0, F(1) = 1
        return n  # Trường hợp cơ bản
    return fib_recursive(n - 1) + fib_recursive(n - 2)  # Công thức F(n) = F(n-1) + F(n-2)

print(fib_recursive(10))  # Fibonacci thứ 10


# Cách 2: Sử dụng vòng lặp 
def fib_dynamic(n):
    """Tính số Fibonacci thứ n bằng quy hoạch động (DP)."""
    if n <= 1: # Trường hợp cơ bản F(0) = 0, F(1) = 1
        return n 
    dp = [0, 1]  # Lưu kết quả Fibonacci trước đó
    for i in range(2, n + 1): # Tính F(n) từ F(2) đến F(n)
        dp.append(dp[-1] + dp[-2])  # Công thức F(n) = F(n-1) + F(n-2)
    return dp[n]

print(fib_dynamic(10))  # Fibonacci thứ 10
